### DISCUSSION TITLE
New pagination bug in GitHub GraphQL: hasNextPage cannot be trusted

### DISCUSSION THREAD

**Discussion by simonw** [community/community#30687] [Apps, API and Webhooks] [answered]
I just started seeing a weird bug in pagination in the GitHub GraphQL API. Consider the following query:

```graphql
{
  viewer {
    repositories(first: 10, privacy: PUBLIC) {
      totalCount
      pageInfo {
        hasNextPage
        endCursor
      }
      nodes {
        name
      }
    }
  }
}
```
Running this against https://docs.github.com/en/graphql/overview/explorer for me returns:

```json
{
  "data": {
    "viewer": {
      "repositories": {
        "totalCount": 568,
        "pageInfo": {
          "hasNextPage": false,
          "endCursor": "Y3Vyc29yOnYyOpHOAALtCQ=="
        },
        "nodes": [
          {
            "name": "django-debug-toolbar"
          },
          {
            "name": "simonw.github.com"
          },
          {
            "name": "ratelimitcache"
          },
          {
            "name": "api-explorer"
          },
          {
            "name": "python-guardianapi"
          },
          {
            "name": "tweetersation"
          },
          {
            "name": "qypeexplorer"
          },
          {
            "name": "getlatlon.com"
          },
          {
            "name": "django-shorturls"
          },
          {
            "name": "soupselect"
          }
        ]
      }
    }
  }
}
```
Note that it shows a totalCount of 568 - which is correct - but despite only listing the first 10 records it returns `"hasNextPage": false`.

This should be `"hasNextPage": true`.

This was working fine a few hours ago - I spotted the bug because my Actions workflow that updates my personal profile on https://github.com/simonw suddenly deleted a bunch of my project listings: https://github.com/simonw/simonw/commit/1ed413252bac947dd8aa9aef15d5ecd5989a7c46

More notes here:
- https://github.com/simonw/simonw/issues/6

**Comment by simonw** [comment]
I found a workaround: don't trust the `hasNextPage` boolean and instead check to see if `endCursor` is `null` in order to terminate pagination.

That seems to have fixed the bug I was seeing.

**Reply by clayne** [reply]
Also seeing this and had to do the exact same thing. Still seems like a bug - and depending on what people are doing with it, could potentially have some fairly fail-dangerous results.

**Reply by UnAfraid** [reply]
If you add `orderBy: {field: CREATED_AT, direction: ASC}` hasNextPage seems to work as expected

**Reply by timrogers** [reply]
@simonw Thanks for flagging this, and apologies for the less-than-positive API experience here! 🙏🏻  I'm the Product Manager for the GitHub API, and I'd love to look into it.

Are you only seeing this behaviour when listing `repositories`, or is it affecting pagination more generally?

**Reply by simonw** [reply]
I've only seen it personally on that `viewer { repositories` edge but I haven't checked for it anywhere else yet.

**Reply by rodgomes** [reply]
I noticed the same, and it seems the workaround doesn't work in my case (it's a search in my org, which contains more than 6k repositories). After a few pages, I also get a null endCursor

**Reply by timrogers** [reply]
@rodgomes I believe this issue should be fixed now. Are you still seeing that behaviour today? If so, would you be able to share your query? 

Feel free to send it to the @github.com email on my profile if there’s anything you want to keep private.

**Reply by rodgomes** [reply]
Hi @timrogers, 
 
I still see the same behavior with this query
```
query RepositorySearch($searchQuery: String!, $cursor: String) {
  search(query: $searchQuery, type: REPOSITORY, first: 100, after: $cursor) {
    
    pageInfo {
      hasNextPage
      endCursor
    }
    repositoryCount
    nodes {
      ... on Repository {
        
        __typename
        id
        isArchived
        name
        nameWithOwner
        url
        description
        visibility
        defaultBranchRef {
          name
        }
        catalogFile: object(expression: "HEAD: .my-config.yaml") {
          ... on Blob {
            text
          }
        }
      }
    }
  }
  rateLimit {
    cost
    remaining
    used
    limit
  }
  
}
```
The variables being only my org for this test, after a few pages I get 
```
...
"search": {
      "pageInfo": {
        "hasNextPage": false,
        "endCursor": "Y3Vyc29yOjEwMDA="
      },
...
```

**Reply by Roohn** [reply]
Correct me if I'm wrong but after some investigation it seems that this last problem is because the search api only returns max 1000 items.

**Reply by timrogers** [reply]
Yep - @Roohn is right. The `search` query only returns a maximum of 1,000 results, even if there are more matches. I'm going to update the docs to clarify that.

**Reply by clayne** [reply]
> Yep - @Roohn is right. The `search` query only returns a maximum of 1,000 results, even if there are more matches. I'm going to update the docs to clarify that.

Where is `RepositorySearch` even documented? I was under the impression it was REST only as I couldn't find a similar query in the GQL docs at the time.

**Reply by timrogers** [reply]
We have a `search` query which allows you to search repos, amongst other things: https://docs.github.com/en/graphql/reference/queries#search

**Reply by clayne** [reply]
> We have a `search` query which allows you to search repos, amongst other things: https://docs.github.com/en/graphql/reference/queries#search

Ah yes, sorry my bad - I see now the actual `search()` query. However, this is still at a disadvantage to the REST based repo search as one cannot specify file or language types, etc. right?

**Reply by timrogers** [reply]
I think should be able to put that kind of thing in the query - but depends exactly what you’re trying to do, I guess!

**Reply by rodgomes** [reply]
Good catch @Roohn and thanks for clarifying @timrogers, I was not aware of the search result limitations.

**Comment by timrogers** [comment, marked as answer]
@simonw Thanks for taking the time to report this bug! It was fixed about a week ago, but I forgot to jump back to this thread and let you know 😓
